lib/actions/ssh-direct.ts

```diff
@@ -0,0 +1,245 @@
+limitations under the License.
+*/
+
+import * as sdk from 'balena-sdk';
+import * as Bluebird from 'bluebird';
+import * as globalTunnel from 'global-tunnel-ng';
+import * as _ from 'lodash';
+import * as net from 'net';
+import * as ssh2 from 'ssh2';
+import * as stream from 'stream';
+
+const balena = sdk.fromSharedOptions();
+
+const tunnelViaSocket = (
+	socket: net.Socket,
+	host: string,
+	port: number,
+	proxyAuth?: Buffer,
+): Bluebird<net.Socket> =>
+	new Bluebird((resolve, reject) => {
+		let tunnelProxyResponse = '';
+		socket.write(`CONNECT ${host}:${port} HTTP/1.0\r\n`);
+		if (proxyAuth != null) {
+			socket.write(
+				`Proxy-Authorization: Basic ${proxyAuth.toString('base64')}\r\n`,
+			);
+		}
+		socket.write('\r\n');
+
+		const earlyEnd = () => {
+			console.error(
+				`Could not connect to ${host}:${port} tunneling socket closed prematurely.`,
+			);
+			reject(
+				new Error(
+					`Could not connect to ${host}:${port} tunneling socket closed prematurely.`,
+				),
+			);
+		};
+		const earlyError = (err: Error) => {
+			let errMsg = 'Could not connect to VPN tunnel';
+			if (err != null && err.message) {
+				errMsg += `: ${err.message}`;
+			}
+			console.error(errMsg);
+			reject(new Error(errMsg));
+		};
+		const proxyData = (chunk: Buffer) => {
+			if (chunk != null) {
+				tunnelProxyResponse += chunk.toString();
+			}
+
+			// read 'data' chunks until full HTTP status line has been read
+			if (!_.includes(tunnelProxyResponse, '\r\n\r\n')) {
+				return;
+			}
+			socket.removeListener('data', proxyData);
+			socket.removeListener('end', earlyEnd);
+			socket.removeListener('error', earlyError);
+
+			// RFC2616: Status-Line = HTTP-Version SP Status-Code SP Reason-Phrase CRLF
+			let httpStatusLine = tunnelProxyResponse.split('\r\n')[0];
+			const httpStatusCode = parseInt(httpStatusLine.split(' ')[1], 10);
+
+			// translate "Proxy-authorization required" to more user-friendly "Forbidden"
+			if (httpStatusCode === 407) {
+				httpStatusLine = 'HTTP/1.0 403 Forbidden';
+			}
+
+			if (httpStatusCode !== 200) {
+				console.error(
+					`Could not connect to ${host}:${port} - ${httpStatusLine}`,
+				);
+				return reject(
+					new Error(`Could not connect to ${host}:${port} - ${httpStatusLine}`),
+				);
+			}
+
+			// one proxied socket, ready to go!
+			resolve(socket);
+		};
+
+		socket
+			.on('end', earlyEnd)
+			.on('error', earlyError)
+			.on('data', proxyData);
+	});
+
+Bluebird.promisifyAll(ssh2.Client);
+
+interface Ssh2ClientAsync extends ssh2.Client {
+	execAsync(
+		cmd: string,
+		options?: ssh2.ExecOptions,
+	): Bluebird<ssh2.ClientChannel>;
+}
+
+const createSSHClient = (socket: stream.Duplex): Bluebird<Ssh2ClientAsync> =>
+	new Bluebird((resolve, reject) => {
+		const client = new ssh2.Client() as Ssh2ClientAsync;
+		return client
+			.on('ready', () => {
+				resolve(client);
+			})
+			.on('error', err => {
+				const errSource = err && err.level ? 'ssh client socket' : 'ssh client';
```

Comment:
Trust your testing rather than my reading, but my reading of the docs suggests that `err.level` would contain one two values: `'client-socket'` or `'client-ssh'`, rather than being set for the former and unset for the latter, in which case the code doesn't look right.
> error(< Error >err) - An error occurred. A 'level' property indicates 'client-socket' for socket-level errors and 'client-ssh' for SSH disconnection messages. In the case of 'client-ssh' messages, there may be a 'description' property that provides more detail.